coordinator/coordinator.go

```diff
@@ -35,34 +36,81 @@ func (c *Coordinator) Run() error {
+	defer c.monitor.Stop()
+
+	var lastActionT time.Time
+	var lastAlertT time.Time
+	var supportedUsers int
+
+	// For now we are keeping all these values constant but in the future they
+	// might change based on the state of the feedback loop.
+	// Ideally we want the value of users to increment/decrement to react
+	// to the speed at which metrics are changing.
+
+	// The value of users to be incremented at each iteration.
+	// It should be proportional to the maximum number of users expected to test.
+	const incValue = 8
+	// The value of users to be decremented at each iteration.
+	// It should be proportional to the maximum number of users expected to test.
+	const decValue = 8
+	// The timespan to wait after a performance degradation alert before
+	// incrementing or decrementing users again.
+	const restTime = 10 * time.Second
 
 	for {
-		status := c.cluster.Status()
-		mlog.Info("coordinator: cluster status:", mlog.Int("active_users", status.ActiveUsers), mlog.Int64("errors", status.NumErrors))
-
-		if status.ActiveUsers < c.config.ClusterConfig.MaxActiveUsers {
-			// TODO: make the choice of this value a bit smarter.
-			inc := 10
-			diff := c.config.ClusterConfig.MaxActiveUsers - status.ActiveUsers
-			if diff < inc {
-				inc = diff
-			}
-			mlog.Info("coordinator: incrementing active users", mlog.Int("num_users", inc))
-			err := c.cluster.IncrementUsers(inc)
-			if err != nil {
-				mlog.Error("coordinator: failed to increment users", mlog.Err(err))
-			}
-		}
+		var perfStatus performance.Status
 
 		select {
 		case <-interruptChannel:
 			mlog.Info("coordinator: shutting down")
-			c.cluster.Shutdown()
 			return nil
-		case <-time.After(1 * time.Second):
+		case perfStatus = <-monitorChan:
+		}
+
+		if perfStatus.Alert {
+			lastAlertT = time.Now()
 		}
 
-		// TODO: implement performance monitoring and act on them to complete feedback loop.
+		status := c.cluster.Status()
+		mlog.Debug("coordinator: cluster status:", mlog.Int("active_users", status.ActiveUsers), mlog.Int64("errors", status.NumErrors))
+
+		// supportedUsers should be estimated in a more clever way in the future.
+		// For now we say that the supported number of users is the number of active users that ran
+		// for the defined timespan without causing any performance degradation alert.
+		if !lastAlertT.IsZero() && !perfStatus.Alert && hasPassed(lastAlertT, restTime) && hasPassed(lastActionT, restTime) {
+			supportedUsers = status.ActiveUsers
+		}
+
+		mlog.Debug("coordinator: supported users", mlog.Int("supported_users", supportedUsers))
+
+		// We give the feedback loop some rest time in case of performance
+		// degradation alerts. We want metrics to stabilize before incrementing/decrementing users again.
+		if lastAlertT.IsZero() || lastActionT.IsZero() || hasPassed(lastActionT, restTime) {
+			if perfStatus.Alert {
+				if err := c.cluster.DecrementUsers(decValue); err != nil {
+					mlog.Error("coordinator: failed to decrement users", mlog.Err(err))
```

Comment:
we are not doing anything if we failed to reduce the number of users? maybe delay and retry (if the server is choked)

Reply:
This falls into the "keep it dead simple" part. If we fail here and performance don't improve we'll just wait and try again anyway so I was hesitant to introduce more logic.